{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RTR Connectivity Graph\n",
    "\n",
    "- Index pruned trends to pruned keywords\n",
    "- Build graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Next steps:\n",
    "\n",
    "- Make Network With Pruned Trends"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create new structure with: \n",
    "- keywords from given abstract\n",
    "- trending words from given abstract\n",
    "\n",
    "TODO: include pubmed link to original abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SETUP\n",
    "%run rtr.ipynb\n",
    "import operator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create data structure to hold trend and associated keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "class Trend:\n",
    "    def __init__(self, trend):\n",
    "        self.trend = trend\n",
    "        #using a set doesn't allow duplicates\n",
    "        #self.keywords = set()\n",
    "        self.keywords = Counter()\n",
    "    def add_keyword(self, keyword):\n",
    "        #self.keywords.add(keyword)\n",
    "        self.keywords.update({keyword:1})\n",
    "    def as_dict(self):\n",
    "        return {self.trend:self.keywords}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pair_trends_keywords(df, window_records, n_trends, n_keywords, from_sql = False):\n",
    "    df_pruned = df.iloc[:n_trends,:]\n",
    "    trends_list = []\n",
    "    for entry in df_pruned.key:\n",
    "        trend_str = ' '.join(entry)\n",
    "        trend = Trend(trend_str)\n",
    "        trends_list.append(trend)\n",
    "    # get filtered keywords\n",
    "    top_keywords = get_top_keywords(n_keywords)\n",
    "    top_keywords_text = [keyword[0] for keyword in top_keywords]\n",
    "    #print(top_keywords_text)\n",
    "    # TRIPLE LOOP - sure to be a bottleneck\n",
    "    for trend in trends_list:\n",
    "        for window in window_records:\n",
    "            # direct method only --\n",
    "            # update to include sql compatibility\n",
    "            # when sql is working\n",
    "            for abstract_record in window:\n",
    "                if from_sql:\n",
    "                    abstract = abstract_record[0]\n",
    "                else:\n",
    "                    abstract = abstract_record['Abstract'][0]\n",
    "                if trend.trend in abstract:\n",
    "                    if from_sql:\n",
    "                        keywords = str.split(abstract_record[2], ',')\n",
    "                    else:\n",
    "                        keywords = abstract_record['Keywords']\n",
    "                    for keyword in keywords:\n",
    "                        if keyword in top_keywords_text:\n",
    "                            #print(\"matching trend\", trend.trend, \"and keyword\", keyword)\n",
    "                            trend.add_keyword(keyword)\n",
    "                #else:\n",
    "                    #print(\"not in abstract\")\n",
    "    return trends_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "def unpickle_data():\n",
    "    with open(\"df.txt\", \"rb\") as fp:   # Unpickling\n",
    "        df = pickle.load(fp)\n",
    "    with open(\"window_records.txt\", \"rb\") as fp:   # Unpickling\n",
    "        window_records = pickle.load(fp)\n",
    "    return df, window_records\n",
    "df, window_records = unpickle_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating engine:\n",
      "-  postgresql://ericbarnhill:carter0109@localhost/rtr_db\n",
      "-  ['rtr_abstracts', 'rtr_keywords']\n",
      "creating engine:\n",
      "-  postgresql://ericbarnhill:carter0109@localhost/rtr_db\n",
      "-  ['rtr_abstracts', 'rtr_keywords']\n",
      "creating engine:\n",
      "-  postgresql://ericbarnhill:carter0109@localhost/rtr_db\n",
      "-  ['rtr_abstracts', 'rtr_keywords']\n"
     ]
    }
   ],
   "source": [
    "n_trends = 1000\n",
    "n_keywords = 1000\n",
    "trends_list = []\n",
    "for sub_df in df:\n",
    "    trends_list.append(pair_trends_keywords(sub_df, window_records, n_trends, n_keywords, from_sql = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'alone combination': Counter()}\n",
      "{'ligament acl': Counter()}\n",
      "{'disease characterized': Counter({'tomography': 6, 'carcinoma': 4, 'pneumonia': 2})}\n",
      "{'tomography used': Counter()}\n",
      "{'tested whether': Counter({'microscopy': 4, 'carcinoma': 2})}\n",
      "{'patients male': Counter()}\n",
      "{'breast imaging': Counter({'imaging': 16, 'tomography': 6, 'diagnosis': 5, 'radiology': 4})}\n",
      "{'patients suspected': Counter({'carcinoma': 4})}\n",
      "{'widely available': Counter({'tomography': 14, 'radiography': 4, 'imaging': 4, 'gene expression regulation': 4})}\n",
      "{'acoustic radiation': Counter({'diagnosis': 5, 'imaging': 3, 'ultrasonography': 2, 'carcinoma': 2})}\n",
      "{'mitral regurgitation': Counter({'echocardiography': 11, 'hypertension': 8, 'osteoporosis': 2, 'diagnosis': 2})}\n",
      "{'body surface': Counter({'tomography': 7, 'imaging': 4, 'carcinoma': 2})}\n",
      "{'investigate potential': Counter()}\n",
      "{'much higher': Counter({'tomography': 8, 'imaging': 8, 'carcinoma': 4, 'rats': 2})}\n",
      "{'vessel disease': Counter({'osteoporosis': 8, 'tomography': 2, 'aged': 2})}\n",
      "{'recurrence rate': Counter({'carcinoma': 11, 'tomography': 6, 'imaging': 5, 'radiography': 2, 'lasers': 2, 'ultrasonography': 2})}\n",
      "{'patients women': Counter()}\n",
      "{'aortic regurgitation': Counter({'echocardiography': 4, 'imaging': 2})}\n",
      "{'ventricular hypertrophy': Counter({'rats': 8, 'echocardiography': 8, 'tomography': 4, 'magnetic resonance imaging': 2, 'electrocardiography': 2, 'imaging': 2, 'heart failure': 1})}\n",
      "{'patients ra': Counter({'tomography': 6, 'mitochondria': 2, 'antibodies': 2})}\n",
      "{'amino acid': Counter({'microscopy': 11, 'tomography': 5, 'dna': 4, 'rats': 2, 'multiple sclerosis': 2, 'mice': 2})}\n",
      "{'prospective observational': Counter({'ultrasonography': 14, 'imaging': 4, 'echocardiography': 2})}\n",
      "{'radiation force': Counter({'imaging': 5, 'diagnosis': 5, 'ultrasonography': 2, 'carcinoma': 2})}\n",
      "{'records patients': Counter()}\n",
      "{'investigated effects': Counter()}\n",
      "{'angiography performed': Counter({'tomography': 4})}\n",
      "{'unknown whether': Counter({'rats': 5, 'imaging': 4})}\n",
      "{'frontal cortex': Counter({'rats': 23, 'tomography': 8, 'mice': 4, 'imaging': 3, 'diagnosis': 2})}\n",
      "{'revealed presence': Counter({'tomography': 2})}\n",
      "{'plays crucial': Counter()}\n",
      "{'right atrium': Counter({'mice': 8, 'ultrasonography': 8, 'imaging': 6, 'biopsy': 6, 'hypertension': 4, 'tomography': 2, 'echocardiography': 2, 'carcinoma': 1})}\n",
      "{'cell viability': Counter({'rats': 13, 'mice': 2, 'imaging': 2, 'carcinoma': 1, 'gene expression regulation': 1})}\n",
      "{'patients experienced': Counter({'echocardiography': 6, 'aged': 4, 'tomography': 4, 'diagnosis': 2})}\n",
      "{'endovascular treatment': Counter({'imaging': 9, 'aneurysm': 8, 'rats': 4, 'tomography': 4})}\n",
      "{'staphylococcus aureus': Counter({'rats': 13, 'aneurysm': 4, 'tomography': 2})}\n",
      "{'even though': Counter({'ultrasonography': 10, 'imaging': 6, 'rats': 5, 'radiography': 2, 'dna': 2, 'intensive care units': 1, 'mice': 1})}\n",
      "{'anterior posterior': Counter()}\n",
      "{'patients multiple': Counter()}\n",
      "{'transthoracic echocardiography': Counter({'echocardiography': 23, 'tomography': 14, 'imaging': 6, 'heart failure': 5, 'hypertension': 4, 'magnetic resonance imaging': 2, 'electrocardiography': 2, 'carcinoma': 1})}\n",
      "{'data regarding': Counter({'tomography': 6, 'echocardiography': 6, 'dna': 2, 'ultrasonography': 2, 'cartilage': 2})}\n",
      "{'real data': Counter({'imaging': 13})}\n",
      "{'inversely associated': Counter({'mice': 4, 'cartilage': 4})}\n",
      "{'left ventricular': Counter({'echocardiography': 117, 'tomography': 44, 'ultrasonography': 30, 'imaging': 26, 'rats': 21, 'mice': 12, 'magnetic resonance imaging': 11, 'heart failure': 9, 'aneurysm': 8, 'electrocardiography': 6, 'hypertension': 4, 'pregnancy': 4, 'cholesterol': 2, 'swine': 2})}\n",
      "{'groups however': Counter()}\n",
      "{'patients including': Counter({'tomography': 2, 'carcinoma': 2})}\n",
      "{'long term': Counter({'tomography': 4, 'aged': 4, 'mice': 4, 'dna': 2, 'imaging': 1, 'rats': 1})}\n",
      "{'epithelial cell': Counter({'mice': 25, 'microscopy': 13, 'imaging': 4, 'rats': 4, 'carcinoma': 3, 'diagnosis': 2, 'dna': 2})}\n",
      "{'spearman rank': Counter({'imaging': 20, 'rats': 4, 'tomography': 4, 'radiotherapy': 2})}\n",
      "{'chronic liver': Counter({'carcinoma': 10, 'diagnosis': 4, 'hypertension': 2, 'tomography': 1})}\n",
      "{'default mode': Counter({'tomography': 10, 'diagnosis': 2, 'rats': 1})}\n",
      "{'used detect': Counter()}\n",
      "{'descriptive statistics': Counter({'mice': 4, 'radiography': 2, 'carcinoma': 2, 'imaging': 2, 'tomography': 1})}\n",
      "{'patients postoperative': Counter({'tomography': 2})}\n",
      "{'analysis results': Counter({'tomography': 6})}\n",
      "{'patients including': Counter({'tomography': 2, 'carcinoma': 2})}\n",
      "{'data regarding': Counter({'tomography': 6, 'echocardiography': 6, 'dna': 2, 'ultrasonography': 2, 'cartilage': 2})}\n",
      "{'resonance energy': Counter({'imaging': 6})}\n",
      "{'safe effective': Counter()}\n",
      "{'records patients': Counter()}\n",
      "{'patients male': Counter()}\n",
      "{'trend toward': Counter({'imaging': 9, 'osteoporosis': 8, 'antibodies': 4, 'radiotherapy': 2, 'ultrasonography': 2, 'carcinoma': 2, 'tomography': 1})}\n",
      "{'acute myocardial': Counter({'tomography': 6, 'echocardiography': 6, 'rats': 4, 'mice': 4, 'aneurysm': 4, 'electrocardiography': 3})}\n",
      "{'patients ra': Counter({'tomography': 6, 'mitochondria': 2, 'antibodies': 2})}\n",
      "{'patients demonstrated': Counter({'echocardiography': 6, 'tomography': 2})}\n",
      "{'superior inferior': Counter()}\n",
      "{'patients early': Counter()}\n",
      "{'boys girls': Counter()}\n",
      "{'cerebral perfusion': Counter({'imaging': 7, 'tomography': 6, 'infant': 4, 'aged': 2, 'intensive care units': 1})}\n",
      "{'common iliac': Counter({'tomography': 2, 'aorta': 1})}\n",
      "{'de novo': Counter({'microscopy': 8, 'imaging': 6, 'tomography': 4, 'mice': 4, 'mutation': 4, 'ultrasonography': 3, 'diagnosis': 1})}\n",
      "{'showed mean': Counter({'imaging': 2})}\n",
      "{'nerve head': Counter({'imaging': 9, 'mice': 1})}\n",
      "{'transthoracic echocardiography': Counter({'echocardiography': 23, 'tomography': 14, 'imaging': 6, 'heart failure': 5, 'hypertension': 4, 'magnetic resonance imaging': 2, 'electrocardiography': 2, 'carcinoma': 1})}\n",
      "{'patients previously': Counter({'tomography': 10, 'diagnosis': 2})}\n",
      "{'myocardial infarction': Counter({'tomography': 36, 'rats': 27, 'echocardiography': 16, 'mice': 15, 'aneurysm': 8, 'electrocardiography': 7, 'cholesterol': 2, 'ultrasonography': 2, 'diagnosis': 2, 'imaging': 2, 'swine': 2, 'magnetic resonance imaging': 2})}\n",
      "{'lesions patients': Counter()}\n",
      "{'older patients': Counter({'diagnosis': 4, 'encephalitis': 2, 'electrocardiography': 2, 'echocardiography': 2})}\n",
      "{'patients identified': Counter({'carcinoma': 2})}\n",
      "{'angiography cta': Counter()}\n",
      "{'years men': Counter()}\n",
      "{'functional brain': Counter({'rats': 8, 'imaging': 4, 'tomography': 4})}\n",
      "{'controls hc': Counter()}\n",
      "{'analogue scale': Counter({'radiography': 4, 'tomography': 3})}\n",
      "{'major complications': Counter({'imaging': 12, 'carcinoma': 11, 'tomography': 10, 'ultrasonography': 2, 'mice': 2})}\n",
      "{'aim compare': Counter()}\n",
      "{'weeks later': Counter({'rats': 8, 'tomography': 1})}\n",
      "{'calculated using': Counter({'imaging': 20, 'tomography': 10, 'echocardiography': 6, 'rats': 5, 'radiography': 4, 'mice': 4, 'ultrasonography': 2, 'carcinoma': 2})}\n",
      "{'soft tissues': Counter({'imaging': 32, 'radiography': 8, 'tomography': 2, 'ultrasonography': 1})}\n",
      "{'obtained results': Counter({'imaging': 12, 'tomography': 2, 'microscopy': 1})}\n",
      "{'prognostic significance': Counter({'tomography': 3, 'heart failure': 2})}\n",
      "{'receiver operator': Counter({'imaging': 2, 'diagnosis': 2})}\n",
      "{'analyzed patients': Counter({'tomography': 2})}\n",
      "{'echocardiographic examination': Counter({'echocardiography': 15, 'tomography': 6, 'ultrasonography': 2})}\n",
      "{'implant placement': Counter({'imaging': 10, 'radiography': 8})}\n",
      "{'lipoprotein cholesterol': Counter({'cholesterol': 4, 'antibodies': 4})}\n",
      "{'bone strength': Counter({'imaging': 6, 'mice': 4, 'osteoporosis': 4, 'rats': 2, 'tomography': 2})}\n",
      "{'aged years': Counter()}\n",
      "{'energy dispersive': Counter({'dna': 2, 'imaging': 2, 'tomography': 2})}\n",
      "{'identify factors': Counter({'tomography': 5, 'diagnosis': 4, 'dna': 2})}\n",
      "{'functional status': Counter({'imaging': 4, 'rats': 2, 'echocardiography': 2})}\n",
      "{'coherence tomography': Counter({'imaging': 56, 'tomography': 17, 'cholesterol': 4, 'rats': 4, 'hypertension': 4, 'cartilage': 4, 'mice': 4, 'aged': 4, 'ultrasonography': 2})}\n",
      "{'lung cancer': Counter({'tomography': 42, 'imaging': 33, 'carcinoma': 12, 'antibodies': 8, 'diagnosis': 5, 'microscopy': 3, 'neoplasms': 2, 'rats': 2, 'biopsy': 2})}\n",
      "{'infarction mi': Counter()}\n",
      "{'mortality morbidity': Counter()}\n",
      "{'young adults': Counter({'tomography': 2, 'imaging': 1})}\n",
      "{'cerebral perfusion': Counter({'imaging': 7, 'tomography': 6, 'infant': 4, 'aged': 2, 'intensive care units': 1})}\n",
      "{'patients receiving': Counter({'imaging': 7, 'carcinoma': 6, 'tomography': 6, 'aged': 2, 'ultrasonography': 1})}\n",
      "{'analysis results': Counter({'tomography': 6})}\n",
      "{'duplex ultrasound': Counter({'tomography': 2, 'ultrasonography': 2})}\n",
      "{'trend toward': Counter({'imaging': 9, 'osteoporosis': 8, 'antibodies': 4, 'radiotherapy': 2, 'ultrasonography': 2, 'carcinoma': 2, 'tomography': 1})}\n",
      "{'functional status': Counter({'imaging': 4, 'rats': 2, 'echocardiography': 2})}\n",
      "{'excluded patients': Counter()}\n",
      "{'fractional flow': Counter({'tomography': 10, 'imaging': 10})}\n",
      "{'records patients': Counter()}\n",
      "{'malignant lesions': Counter({'tomography': 8, 'carcinoma': 8, 'diagnosis': 7, 'imaging': 4})}\n",
      "{'treatment strategy': Counter({'rats': 4, 'diagnosis': 2})}\n",
      "{'late gadolinium': Counter({'magnetic resonance imaging': 11, 'electrocardiography': 6, 'tomography': 5, 'echocardiography': 3, 'diagnosis': 2})}\n",
      "{'aortic aneurysms': Counter({'imaging': 18, 'tomography': 7, 'aneurysm': 6})}\n",
      "{'performed cases': Counter()}\n",
      "{'single institution': Counter({'imaging': 8, 'aneurysm': 3, 'radiotherapy': 2, 'tomography': 2})}\n",
      "{'largely unknown': Counter({'mice': 13, 'rats': 12, 'imaging': 6, 'microscopy': 4, 'tomography': 2, 'spectrum analysis': 1})}\n",
      "{'signs symptoms': Counter()}\n",
      "{'transthoracic echocardiography': Counter({'echocardiography': 23, 'tomography': 14, 'imaging': 6, 'heart failure': 5, 'hypertension': 4, 'magnetic resonance imaging': 2, 'electrocardiography': 2, 'carcinoma': 1})}\n",
      "{'regions including': Counter({'rats': 8, 'tomography': 2})}\n",
      "{'provide important': Counter({'imaging': 5, 'cartilage': 4, 'rats': 4, 'tomography': 2})}\n",
      "{'cognitive deficits': Counter({'rats': 15, 'tomography': 4})}\n",
      "{'vessel diameter': Counter({'imaging': 4, 'tomography': 3})}\n",
      "{'patients early': Counter()}\n",
      "{'least years': Counter()}\n",
      "{'weeks later': Counter({'rats': 8, 'tomography': 1})}\n",
      "{'renal artery': Counter({'echocardiography': 6, 'ultrasonography': 6, 'carcinoma': 4, 'tomography': 3, 'rats': 2})}\n",
      "{'patients chronic': Counter()}\n",
      "{'rv function': Counter({'echocardiography': 5, 'rats': 4, 'magnetic resonance imaging': 2, 'hypertension': 2})}\n",
      "{'ventricular rv': Counter()}\n",
      "{'women aged': Counter({'carcinoma': 4, 'osteoporosis': 2})}\n",
      "{'adverse cardiovascular': Counter({'tomography': 4, 'magnetic resonance imaging': 2, 'echocardiography': 2})}\n",
      "{'patients presented': Counter({'imaging': 9, 'tomography': 7, 'diagnosis': 6, 'aneurysm': 2, 'cartilage': 1})}\n",
      "{'analyses used': Counter()}\n",
      "{'acl reconstruction': Counter({'cartilage': 16, 'imaging': 8})}\n",
      "{'tomography angiography': Counter({'tomography': 35, 'imaging': 24, 'diagnosis': 2, 'aneurysm': 2})}\n",
      "{'chest wall': Counter({'imaging': 4, 'hypertension': 4, 'tomography': 2, 'ultrasonography': 1, 'carcinoma': 1})}\n",
      "{'microcomputed tomography': Counter({'mice': 12, 'rats': 4, 'imaging': 3, 'tomography': 2})}\n",
      "{'membrane potential': Counter({'rats': 10, 'imaging': 6, 'microscopy': 2})}\n",
      "{'ischemic stroke': Counter({'rats': 21, 'tomography': 5, 'aneurysm': 4, 'aged': 2, 'imaging': 2, 'electrocardiography': 1, 'diagnosis': 1})}\n",
      "{'analyzed patients': Counter({'tomography': 2})}\n",
      "{'patients control': Counter()}\n",
      "{'stent grafts': Counter()}\n",
      "{'retrospective chart': Counter({'imaging': 4, 'tomography': 2, 'biopsy': 2, 'diagnosis': 1})}\n",
      "{'injury tbi': Counter()}\n",
      "{'strain gls': Counter()}\n",
      "{'fiber layer': Counter({'imaging': 7, 'tomography': 4})}\n",
      "{'studies confirmed': Counter({'rats': 4, 'radiotherapy': 1})}\n",
      "{'nerve head': Counter({'imaging': 9, 'mice': 1})}\n"
     ]
    }
   ],
   "source": [
    "def print_trend_dict(trends_list, N=50):\n",
    "    n = 0\n",
    "    for trend in trends_list:\n",
    "        print(trend.as_dict())\n",
    "        n += 1\n",
    "        if n > N:\n",
    "            return\n",
    "for trends_sublist in trends_list:\n",
    "    print_trend_dict(trends_sublist)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tomography , photon emission , weight:  243\n",
      "tomography , single photon , weight:  103\n",
      "tomography , coronary artery , weight:  111\n",
      "imaging , coherence tomography , weight:  56\n",
      "imaging , optical coherence , weight:  58\n",
      "imaging , healthy volunteers , weight:  67\n",
      "imaging , soft tissue , weight:  96\n",
      "echocardiography , left ventricular , weight:  117\n",
      "echocardiography , doppler echocardiography , weight:  57\n",
      "imaging , coherence tomography , weight:  56\n",
      "imaging , optical coherence , weight:  58\n",
      "imaging , soft tissue , weight:  96\n",
      "imaging , determine whether , weight:  53\n",
      "tomography , coronary artery , weight:  111\n",
      "tomography , artery disease , weight:  80\n",
      "echocardiography , left ventricular , weight:  117\n",
      "echocardiography , heart failure , weight:  67\n",
      "cartilage , articular cartilage , weight:  95\n",
      "tomography , artery disease , weight:  80\n",
      "tomography , imaging technique , weight:  53\n",
      "tomography , coronary artery , weight:  111\n",
      "imaging , coherence tomography , weight:  56\n",
      "imaging , optical coherence , weight:  58\n",
      "imaging , imaging technique , weight:  65\n",
      "imaging , electron microscopy , weight:  52\n",
      "imaging , data set , weight:  57\n",
      "rats , stem cell , weight:  54\n",
      "rats , electron microscopy , weight:  93\n",
      "microscopy , electron microscopy , weight:  88\n",
      "cartilage , articular cartilage , weight:  95\n"
     ]
    }
   ],
   "source": [
    "def populate_graph(trends_list):\n",
    "    G = nx.Graph()\n",
    "    for trend in trends_list:\n",
    "        node1 = trend.trend\n",
    "        for key, item in trend.keywords.items():\n",
    "            node2 = key\n",
    "            #print(node1, \"|\", node2, \"|\", item)\n",
    "            if G.has_edge(node1, node2):\n",
    "                print(\"increasing weight by\", item)\n",
    "                G[node1][node2]['weight'] += item\n",
    "            else:\n",
    "                # new edge. add with weight=1\n",
    "                G.add_node(node1, is_key = False)\n",
    "                G.add_node(node2, is_key = True)\n",
    "                G.add_edge(node1, node2, weight=item)\n",
    "    for u, v, d in G.edges(data=True):\n",
    "        weight = d['weight']\n",
    "        if weight > 50:\n",
    "            print(u,\",\" , v, \",\", \"weight: \", d['weight'])\n",
    "    return G\n",
    "G = []\n",
    "for trends_sublist in trends_list:\n",
    "    G.append(populate_graph(trends_sublist))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find ten most heavily weighted edges of the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['soft tissue 96', 'healthy volunteers 67', 'optical coherence 58', 'coherence tomography 56', 'results demonstrate 48', 'mr image 46', 'important role 45', 'treatment planning 45', 'image acquisition 44', 'blood flow 43']\n",
      "---\n",
      "['soft tissue 96', 'optical coherence 58', 'coherence tomography 56', 'determine whether 53', 'navigation system 45', 'image analysis 42', 'ex vivo 37', 'coronary artery 36', 'tomography scan 36', 'see text 36']\n",
      "---\n",
      "['imaging technique 65', 'optical coherence 58', 'data set 57', 'coherence tomography 56', 'electron microscopy 52', 'blood flow 43', 'ex vivo 37', 'coronary artery 36', 'mean absolute 36', 'tomography scan 36']\n",
      "---\n"
     ]
    }
   ],
   "source": [
    "def top_N_trends(G, keyword, n=10):\n",
    "    node_edges = G.edges(keyword.lower())\n",
    "    edges_dict = {}\n",
    "    for edge in node_edges:\n",
    "        key = edge[1]\n",
    "        value = G[edge[0]][edge[1]]['weight']\n",
    "        edges_dict.update({key:value})\n",
    "    index = 1\n",
    "    trends = []\n",
    "    for key, value in sorted(edges_dict.items(), key=operator.itemgetter(1), reverse=True):\n",
    "        trends.append(' '.join([key, str(value)]))\n",
    "        index += 1\n",
    "        if index > n:\n",
    "            break\n",
    "    return trends\n",
    "for G_sub in G:\n",
    "    trends = top_N_trends(G_sub, 'imaging')\n",
    "    print(trends)\n",
    "    print('---')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find which keywords made the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('imaging', 4254)\n",
      "('tomography', 3857)\n",
      "('rats', 1960)\n",
      "('ultrasonography', 1364)\n",
      "('echocardiography', 1149)\n",
      "('carcinoma', 880)\n",
      "('mice', 856)\n",
      "('diagnosis', 521)\n",
      "('microscopy', 430)\n",
      "('cartilage', 323)\n",
      "('radiography', 268)\n",
      "('hypertension', 218)\n",
      "('osteoporosis', 210)\n",
      "('aneurysm', 191)\n",
      "('dna', 175)\n",
      "('aged', 142)\n",
      "('cholesterol', 118)\n",
      "('magnetic resonance imaging', 108)\n",
      "('electrocardiography', 103)\n",
      "('evoked potentials', 103)\n",
      "('biopsy', 97)\n",
      "---\n",
      "('imaging', 3826)\n",
      "('tomography', 3605)\n",
      "('rats', 1637)\n",
      "('ultrasonography', 1175)\n",
      "('carcinoma', 973)\n",
      "('echocardiography', 958)\n",
      "('mice', 821)\n",
      "('diagnosis', 554)\n",
      "('cartilage', 499)\n",
      "('microscopy', 357)\n",
      "('radiography', 274)\n",
      "('aneurysm', 246)\n",
      "('heart failure', 222)\n",
      "('hypertension', 200)\n",
      "('dna', 174)\n",
      "('magnetic resonance imaging', 156)\n",
      "('osteoporosis', 146)\n",
      "('aged', 140)\n",
      "('electrocardiography', 127)\n",
      "('cholesterol', 124)\n",
      "('gene expression regulation', 96)\n",
      "---\n",
      "('imaging', 3542)\n",
      "('tomography', 3441)\n",
      "('rats', 1841)\n",
      "('ultrasonography', 1057)\n",
      "('carcinoma', 933)\n",
      "('mice', 730)\n",
      "('echocardiography', 705)\n",
      "('diagnosis', 505)\n",
      "('microscopy', 480)\n",
      "('cartilage', 454)\n",
      "('radiography', 266)\n",
      "('dna', 246)\n",
      "('aneurysm', 242)\n",
      "('hypertension', 192)\n",
      "('magnetic resonance imaging', 163)\n",
      "('aged', 157)\n",
      "('electrocardiography', 130)\n",
      "('osteoporosis', 128)\n",
      "('cholesterol', 82)\n",
      "('gene expression regulation', 74)\n",
      "('biopsy', 69)\n",
      "---\n"
     ]
    }
   ],
   "source": [
    "def get_hot_keywords(G):\n",
    "    N = 20\n",
    "    keywords =  {}\n",
    "    for node in G.nodes(data=True):\n",
    "        if node[1]['is_key']:\n",
    "            total_wt = 0\n",
    "            for edge in G.edges(node[0], data=True):\n",
    "                total_wt += edge[2]['weight']\n",
    "            keywords.update({node[0]: total_wt})\n",
    "    n = 0\n",
    "    for item in sorted(keywords.items(), key=operator.itemgetter(1), reverse=True):\n",
    "        print(item)\n",
    "        n += 1\n",
    "        if n > N:\n",
    "            break\n",
    "for G_sub in G:\n",
    "    get_hot_keywords(G_sub)\n",
    "    print('---')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pickle graph\n",
    "def pickle_graph():\n",
    "    try: \n",
    "        with open(\"/home/ericbarnhill/Documents/code/insight/rtr/G.txt\", \"wb\") as fp:\n",
    "            pickle.dump(G, fp)\n",
    "    except:\n",
    "        print('error')\n",
    "pickle_graph()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.exists(\"/home/ericbarnhill/Documents/code/insight/rtr/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating engine:\n",
      "-  postgresql://ericbarnhill:carter0109@localhost/rtr_db\n",
      "-  ['rtr_abstracts', 'rtr_keywords']\n"
     ]
    }
   ],
   "source": [
    "keywords = get_top_keywords(1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
